### PR TITLE
Add support for external image registries

### PR DESCRIPTION
**Pull request by jameshalsall**
External image registries can be defined like this:

```yaml
docker:
    external_image_registries:
      - url: quay.io
        username: foo
        password: = decrypt('...')
```

TODO:
- [x] get feedback
- [x] update documentation and write how-to guide

### CODE REVIEW

File: harness/config/external-images.yml [hunk 1]
```diff
@@ -36,11 +36,27 @@ command('external-images config [--skip-exists]'):
...
+      shell_exec("echo $pass | docker login --username=$user --password-stdin $url");
+    }
+
+    $confArgs = [];
+    if (strlen(@$env['CI']) > 0 || strlen(@$env['BUILD_ID']) > 0) {
+      $confArgs[] = "--skip-exists";
+    }
+    $confArgs = escapeshellarg(join(' ', $confArgs));
+
+    shell_exec("ws external-images config $confArgs | docker-compose -f - pull");
```
Comment (andytson-inviqa) [inline harness/config/external-images.yml:57]: passthru() would be useful here, but it's not like the bash function ws has, so best logging the command run before running it

File: harness/config/external-images.yml [hunk 2]
```diff
@@ -36,11 +36,27 @@ command('external-images config [--skip-exists]'):
...
+command('external-images pull'):
+  env:
+    EXTERNAL_REGISTRIES: = json_encode(@('docker.external_image_registries'))
+  exec: |
+    #!php
+    foreach (json_decode($env['EXTERNAL_REGISTRIES'], true) as $registry) {
+      $pass = escapeshellarg($registry['password']);
+      $user = escapeshellarg($registry['username']);
+      $url = escapeshellarg($registry['url']);
+      shell_exec("echo $pass | docker login --username=$user --password-stdin $url");
```
Comment (andytson-inviqa) [inline harness/config/external-images.yml:48]: less useful to log the command here, but passthru would be useful for displayign errors, also php doesn't exception on failures here

File: harness/config/external-images.yml (hunk 1 above)
Comment (andytson-inviqa) [reply harness/config/external-images.yml:57]: also php doesn't exception on errors here

File: harness/config/external-images.yml [hunk 3]
```diff
@@ -36,11 +36,33 @@ command('external-images config [--skip-exists]'):
...
+      $command = "docker login --username=$user --password-stdin $url";
+      echo $command . PHP_EOL;
+      passthru("echo $pass | $command");
+    }
+
+    $confArgs = [];
+    if (strlen(@$env['CI']) > 0 || strlen(@$env['BUILD_ID']) > 0) {
+      $confArgs[] = "--skip-exists";
+    }
+    $confArgs = escapeshellarg(join(' ', $confArgs));
```
Comment (andytson-inviqa) [inline harness/config/external-images.yml:57]:
```suggestion
    $confArgs = join(' ', array_map('escapeshellarg',$confArgs));
```
which is more like the `$(printf '%q ' "${CONF_ARGS[@]}")`

Comment (jameshalsall) [reply harness/config/external-images.yml:57]: niiiice

File: docs/how-to-guides/use-private-external-images.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+   `tools/workspace/_twig/docker-compose.yml/service/my-service.yml.twig`:
+   ```yaml
+     my-service:
+      image: quay.io/my_images/foo:latest
+      environment:
+        FOO: "bar"
+      networks:
+        private: {}
+   ```
+4. Add your  to the `app.services` attribute:
```
Comment (andytson-inviqa) [inline docs/how-to-guides/use-private-external-images.md:27]:
```suggestion
3. Add your  to the `app.services` attribute:
```

**Review by andytson-inviqa — APPROVED** [review]

**Pull request merged**